README.md
Chronic
=======

## DESCRIPTION

[Chronic](https://github.com/mojombo/chronic) is a natural language date/time parser written in pure Ruby.
Chronic L10n aims to add translations to [Chronics](https://github.com/mojombo/chronic) abilities.


## INSTALLATION

*WARNING*: This gem will only work when and **IF** [this fork](https://github.com/luan/chronic) of chronic gets merged and released.
Meanwhile you can install the [chronic](https://github.com/luan/chronic) gem from the forks source to get it going.

### RubyGems

    $ [sudo] gem install chronic-l10n

### GitHub

    $ git clone git://github.com/luan/chronic-l10n.git
    $ cd chronic-l10n && gem build chronic-l10n.gemspec
    $ gem install chronic-l10n-<version>.gem


## USAGE

You can parse strings containing a natural language date using the
`Chronic.parse` method.

    require 'chronic'
    require 'chronic-l10n'

    Time.now   #=> Sun Aug 27 23:18:25 PDT 2006

    Chronic.locale = :'pt-BR'

    Chronic.parse('amanhã')
      #=> Mon Aug 28 12:00:00 PDT 2006

    Chronic.parse('segunda', :context => :past)
      #=> Mon Aug 21 12:00:00 PDT 2006

    Chronic.parse('essa terça 5:00')
      #=> Tue Aug 29 17:00:00 PDT 2006

    Chronic.parse('essa terça 5:00', :ambiguous_time_range => :none)
      #=> Tue Aug 29 05:00:00 PDT 2006

    Chronic.parse('27 de maio', :now => Time.local(2000, 1, 1))
      #=> Sat May 27 12:00:00 PDT 2000

    Chronic.parse('27 de maio', :guess => false)
      #=> Sun May 27 00:00:00 PDT 2007..Mon May 28 00:00:00 PDT 2007
      
    Chronic.parse('6/4/2012', :endian_precedence => :little)
      #=> Fri Apr 06 00:00:00 PDT 2012
      

See `Chronic.parse` for detailed usage instructions.

## CONTRIBUTE

You might like to add your language or improve some already available language, start by forking the repo on GitHub:

https://github.com/luan/chronic-l10n

The best way to get your changes merged back into core is as follows:

1. Clone down your fork
1. Create a thoughtfully named topic branch to contain your change
1. Hack away
1. Add tests and make sure everything still passes by running `rake`
1. Ensure your tests pass in multiple timezones. ie `TZ=utc rake` `TZ=BST rake`
1. If you are adding new functionality, document it in the README
1. Do not change the version number, we will do that on our end
1. If necessary, rebase your commits into logical chunks, without errors
1. Push the branch up to GitHub
1. Send a pull request for your branch


HISTORY.md
# HEAD

* Added PT-BR


test/helper.rb
require 'chronic'
unless defined? Chronic::L10n
  $:.unshift File.expand_path('../../lib', __FILE__)
  require 'chronic-l10n'
end

require 'minitest/autorun'

class TestCase < MiniTest::Unit::TestCase
  def self.test(name, &block)
    define_method("test_#{name.gsub(/\W/, '_')}", &block) if block
  end
end


test/test_parsing_pt_br.rb
# encoding: utf-8
require 'helper'

class TestParsingPtBr < TestCase
  # Wed Aug 16 14:00:00 UTC 2006
  TIME_2006_08_16_14_00_00 = Time.local(2006, 8, 16, 14, 0, 0, 0)

  def setup
    @time_2006_08_16_14_00_00 = TIME_2006_08_16_14_00_00
    @locale_before = Chronic.locale
    Chronic.locale = :'pt-BR'
  end

  def teardown
    Chronic.locale = @locale_before
  end

  def test_handle_basics
    time = parse_now("3 de agosto")
    assert_equal Time.local(2006, 8, 3, 12), time

    time = parse_now("3 de agosto", :context => :past)
    assert_equal Time.local(2006, 8, 3, 12), time

    time = parse_now("20 de agosto")
    assert_equal Time.local(2006, 8, 20, 12), time

    time = parse_now("20 de agosto", :context => :future)
    assert_equal Time.local(2006, 8, 20, 12), time

    time = parse_now("27 de maio")
    assert_equal Time.local(2007, 5, 27, 12), time

    time = parse_now("28 de maio", :context => :past)
    assert_equal Time.local(2006, 5, 28, 12), time

    time = parse_now("28 de maio 5pm", :context => :past)
    assert_equal Time.local(2006, 5, 28, 17), time

    time = parse_now("28 de maio às 17", :context => :future)
    assert_equal Time.local(2007, 5, 28, 17), time

    time = parse_now("28 de maio as 5 da tarde", :context => :past)
    assert_equal Time.local(2006, 5, 28, 17), time

    time = parse_now("28 de maio as 5:32.19pm", :context => :past)
    assert_equal Time.local(2006, 5, 28, 17, 32, 19), time

    time = parse_now("5pm maio 28")
    assert_equal Time.local(2007, 5, 28, 17), time

    time = parse_now("17h, 28 de maio")
    assert_equal Time.local(2007, 5, 28, 17), time

    time = parse_now("5 maio 28", :ambiguous_time_range => :none)
    assert_equal Time.local(2007, 5, 28, 05), time
  end

  def test_textual_forms
    time = parse_now("amanhã")
    assert_equal Time.local(2006, 8, 17, 12), time

    time = parse_now("ontem")
    assert_equal Time.local(2006, 8, 15, 12), time

    time = parse_now("sabado")
    assert_equal Time.local(2006, 8, 19, 12), time

    time = parse_now("vinte e um de agosto")
    assert_equal Time.local(2006, 8, 21, 12), time
  end

  def test_day_portions
    time = parse_now("2 da madrugada")
    assert_equal Time.local(2006, 8, 16, 2), time

    time = parse_now("2 da manha")
    assert_equal Time.local(2006, 8, 16, 2), time

    time = parse_now("2 da tarde")
    assert_equal Time.local(2006, 8, 16, 14), time

    time = parse_now("7 da noite")
    assert_equal Time.local(2006, 8, 16, 19), time
  end

  def test_combinations
    time = parse_now("sabado as 10")
    assert_equal Time.local(2006, 8, 19, 10), time

    time = parse_now("este domingo as 19h")
    assert_equal Time.local(2006, 8, 20, 19), time

    time = parse_now("proxima quarta 1 da manha")
    assert_equal Time.local(2006, 8, 23, 1), time

    time = parse_now("terça passada 5 da tarde")
    assert_equal Time.local(2006, 8, 15, 17), time
  end

  def test_relative_times
    time = parse_now("7 horas depois")
    assert_equal Time.local(2006, 8, 16, 21), time

    time = parse_now("daqui a 7 horas")
    assert_equal Time.local(2006, 8, 16, 21), time

    time = parse_now("5 horas atras")
    assert_equal Time.local(2006, 8, 16, 9), time

    time = parse_now("5 horas antes")
    assert_equal Time.local(2006, 8, 16, 9), time

    time = parse_now("2 horas depois do meio dia")
    assert_equal Time.local(2006, 8, 16, 14), time

    time = parse_now("30 minutos antes da meia noite")
    assert_equal Time.local(2006, 8, 16, 23, 30), time
  end

  def test_relative_dates
    time = parse_now("mes que vem")
    assert_equal Time.local(2006, 9, 16), time

    time = parse_now("proximo mes")
    assert_equal Time.local(2006, 9, 16), time

    time = parse_now("mes passado")
    assert_equal Time.local(2006, 7, 16, 12), time

    time = parse_now("mes anterior")
    assert_equal Time.local(2006, 7, 16, 12), time

    time = parse_now("ultimo mes")
    assert_equal Time.local(2006, 7, 16, 12), time

    time = parse_now("daqui 3 dias")
    assert_equal Time.local(2006, 8, 19, 14), time

    time = parse_now("daqui 3 semanas")
    assert_equal Time.local(2006, 9, 6, 14), time

    time = parse_now("hoje as 9")
    assert_equal Time.local(2006, 8, 16, 9, 00), time

    time = parse_now("agora")
    assert_equal Time.local(2006, 8, 16, 14, 00), time
  end

  private
  def parse_now(string, options={})
    Chronic.parse(string, {:now => TIME_2006_08_16_14_00_00 }.merge(options))
  end
end


lib/chronic-l10n/pt_br.rb
def removeaccents(string)
  accents = {
    'E' => [200,201,202,203],
    'e' => [232,233,234,235],
    'A' => [192,193,194,195,196,197],
    'a' => [224,225,226,227,228,229,230],
    'C' => [199],
    'c' => [231],
    'O' => [210,211,212,213,214,216],
    'o' => [242,243,244,245,246,248],
    'I' => [204,205,206,207],
    'i' => [236,237,238,239],
    'U' => [217,218,219,220],
    'u' => [249,250,251,252],
    'N' => [209],
    'n' => [241],
    'Y' => [221],
    'y' => [253,255],
    'AE' => [306],
    'ae' => [346],
    'OE' => [188],
    'oe' => [189]
  }

  str = String.new(string)
  accents.each do |letter,accents|
    packed = accents.pack('U*')
    rxp = Regexp.new("[#{packed}]", nil)
    str.gsub!(rxp, letter)
  end

  str
end

module Chronic
  module L10n
    PT_BR = {
      :pointer => {
        /\bpassad[oa]\b/ => :past,
        /\b(?:futuro|em|que vem)\b/ => :future,
      },
      :ordinal_regex => /^(\d*)[oa]$/,
      :numerizer => {
        :and => 'e',
        :preprocess => [
          [/ +|([^\d])-([^\d])/, '\1 \2'], # will mutilate hyphenated-words but shouldn't matter for date extraction
          [/e mei[ao]/, 'meia'] # take the 'a' out so it doesn't turn into a 1, save the half for the end
        ],
        :fractional => [
          [/(\d+)(?: |-)*mei[ao]/i, '\1:30']
        ],
        :direct_nums => [
          ['onze', '11'],
          ['doze', '12'],
          ['treze', '13'],
          ['quartorze', '14'],
          ['catorze', '14'],
          ['quinze', '15'],
          ['dezesseis', '16'],
          ['dezeseis', '16'],
          ['dezessete', '17'],
          ['dezesete', '17'],
          ['dezoito', '18'],
          ['dezenove', '19'],
          ['zero', '0'],
          ['um', '1'],
          ['uma', '1'],
          ['dois', '2'],
          ['duas', '2'],
          ['tres', '3'],
          ['quatro', '4'],  # The weird regex is so that it matches four but not fourty
          ['cinco', '5'],
          ['seis', '6'],
          ['sete', '7'],
          ['oito', '8'],
          ['nove', '9'],
          ['dez(\W|$)', '10\1']
        ],
        :ordinals => [
          ['primeiro', '1'],
          ['terceiro', '3'],
          ['quarto', '4'],
          ['quinto', '5'],
          ['sexto', '6'],
          ['setimo', '7'],
          ['oitavo', '8'],
          ['nono', '9'],
          ['decimo', '10']
        ],
        :ten_prefixes => [
          ['vinte', 20],
          ['trinta', 30],
          ['quarenta', 40],
          ['cinquenta', 50],
          ['cincuenta', 50],
          ['sessenta', 60],
          ['setenta', 70],
          ['oitenta', 80],
          ['noventa', 90]
        ],
        :big_prefixes => [
          ['cem', 100],
          ['mil', 1000],
          ['milhao', 1_000_000],
          ['bilhao', 1_000_000_000],
          ['trilhao', 1_000_000_000_000],
        ],
      },

      :repeater => {
        :season_names => {
          /^primaveras?$/ => :spring,
          /^ver(ao|oes)$/ => :summer,
          /^outonos?$/ => :autumn,
          /^invernos?$/ => :winter
        },
        :month_names => {
          /^jan\.?(eiro)?$/ => :january,
          /^fev\.?(ereiro)?$/ => :february,
          /^mar\.?(co)?$/ => :march,
          /^abr\.?(il)?$/ => :april,
          /^mai\.?o?$/ => :may,
          /^jun\.?(ho)?$/ => :june,
          /^jul\.?(ho)?$/ => :july,
          /^ago\.?(sto)?$/ => :august,
          /^set\.?(embro)?$/ => :september,
          /^out\.?(ubro)?$/ => :october,
          /^nov\.?(embro)?$/ => :november,
          /^dez\.?(embro)?$/ => :december
        },
        :day_names => {
          /^seg(unda)?(-feira)?$/ => :monday,
          /^ter(ca)?(-feira)?$/ => :tuesday,
          /^qua(rta)?(-feira)?$/ => :wednesday,
          /^qui(nta)?(-feira)?$/ => :thursday,
          /^sex(ta)?(-feira)?$/ => :friday,
          /^sab(ado)?$/ => :saturday,
          /^dom(ingo)?$/ => :sunday
        },
        :day_portions => {
          /^ams?$/ => :am,
          /^pms?$/ => :pm,
          /^(madrugada|manha)s?$/ => :morning,
          /^tardes?$/ => :afternoon,
          /^noites?$/ => :evening,
          /^noites?$/ => :night
        },
        :units => {
          /^anos?$/ => :year,
          /^estacoes?$/ => :season,
          /^mes(es)?$/ => :month,
          /^quinzenas?$/ => :fortnight,
          /^semanas?$/ => :week,
          /^fi(m|ns) de semanas?$/ => :weekend,
          /^dias? ut(il|eis)$/ => :weekday,
          /^dias?$/ => :day,
          /^hrs?$/ => :hour,
          /^horas?$/ => :hour,
          /^mins?$/ => :minute,
          /^minutos?$/ => :minute,
          /^secs?$/ => :second,
          /^segundos?$/ => :second
        }
      },

      :pre_normalize => {
        :preprocess => proc {|str| removeaccents(str)},
        :pre_numerize => [
          [/\./, ':'],
          [/['"]/, ''],
          [/(.*),(.*)/, '\2 \1'],
          [/^segundo /, '2nd '],
          [/\bsegundo (de|dia|mes|hora|ninuto|segundo)\b/, '2nd \1']
        ],
        :pos_numerize => [
          [/ \-(\d{4})\b/, ' tzminus\1'],
          [/([\/\-\,\@])/, ' \1 '],
          [/(?:^|\s)0(\d+:\d+\s*pm?\b)/, ' \1'],
          [/\bhoje\b/, 'este dia'],
          [/\bamanha\b/, 'proximo dia'],
          [/\bontem\b/, 'ultimo dia'],
          [/\b(\w+) (?:anterior|passad[ao])\b/, 'ultimo \1'],
          [/\b(\w+) (?:futuro|que vem)\b/, 'proximo \1'],
          [/\bmeio[- ]dia\b/, '12:00pm'],
          [/\bmeia[- ]noite\b/, '24:00'],
          [/\bagora\b/, 'este segundo'],
          [/\b(?:da|de) (madrugada|manha)\b/, '\1'],
          [/\b(?:da|de|a) (tarde|noite)\b/, '\1'],
          [/\bhoje a noite\b/, 'esta noite'],
          [/\b\d+:?\d*[ap]\b/,'\0m'],
          [/(\d)([ap]m)\b/, '\1 \2'],
          [/(\d)(?:h|em ponto)\b/, '\1:00'],
          [/\b(?:daqui a|daqui|depois)\b/, 'futuro'],
          [/\b(?:antes|atras)\b/, 'passado'],
          [/\b(\d+) de (\w+)\b/, '\2 \1']
        ]
      },

      :grabber => {
        /ultim[ao]/ => :last,
        /est[ae]/ => :this,
        /proxim[ao]/ => :next
      },

      :token => {
        :comma => /^,$/,
        :at => /^(as|@)$/,
        :in => /^em$/,
        :on => /^em$/
      }
    }
  end
end
